parties/githubclient.go

```diff
@@ -1,18 +1,236 @@
 package parties
 
 import (
+	"bytes"
+	"context"
 	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/bastionzero/openpubkey/gq"
+	"github.com/bastionzero/openpubkey/pktoken"
+	"github.com/bastionzero/openpubkey/util"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/zitadel/oidc/v2/pkg/client"
 )
 
+const githubIssuer = "https://token.actions.githubusercontent.com"
+
 type GithubOp struct {
+	rawTokenRequestURL    string
+	tokenRequestAuthToken string
+}
+
+var _ OpenIdProvider = (*GithubOp)(nil)
+
+func NewGithubOpFromEnvironment() (*GithubOp, error) {
+	tokenURL, err := getEnvVar("ACTIONS_ID_TOKEN_REQUEST_URL")
+	if err != nil {
+		return nil, err
+	}
+	token, err := getEnvVar("ACTIONS_ID_TOKEN_REQUEST_TOKEN")
+	if err != nil {
+		return nil, err
+	}
+
+	return NewGithubOp(tokenURL, token), nil
+}
+
+func getEnvVar(name string) (string, error) {
+	value, ok := os.LookupEnv(name)
+	if !ok {
+		return "", fmt.Errorf("%q environment variable not set", name)
+	}
+	return value, nil
+}
+
+func NewGithubOp(tokenURL string, token string) *GithubOp {
+	return &GithubOp{
+		rawTokenRequestURL:    tokenURL,
+		tokenRequestAuthToken: token,
+	}
+}
+
+func buildTokenURL(rawTokenURL, audience string) (string, error) {
+	parsedURL, err := url.Parse(rawTokenURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse URL: %w", err)
+	}
+
+	if audience == "" {
+		return "", fmt.Errorf("audience is required")
+	}
+
+	query := parsedURL.Query()
+	query.Set("audience", audience)
+	parsedURL.RawQuery = query.Encode()
+	return parsedURL.String(), nil
 }
 
-func (h *GithubOp) RequestTokens(cicHash string) ([]byte, error) {
-	// TODO: Add github action call here
-	return nil, nil
+func (g *GithubOp) PublicKey(idt []byte) (PublicKey, error) {
+	j, err := jws.Parse(idt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JWS: %w", err)
+	}
+	headers := j.Signatures()[0].ProtectedHeaders()
+	alg, kid := headers.Algorithm(), headers.KeyID()
+	if alg != jwa.RS256 {
+		return nil, fmt.Errorf("expected RS256 alg claim, got %s", alg)
+	}
+
+	discConf, err := client.Discover(githubIssuer, http.DefaultClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call OIDC discovery endpoint: %w", err)
+	}
+
+	jwks, err := jwk.Fetch(context.TODO(), discConf.JwksURI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch to JWKS: %w", err)
+	}
+
+	key, ok := jwks.LookupKeyID(kid)
+	if !ok {
+		return nil, fmt.Errorf("key %q isn't in JWKS", kid)
+	}
+	keyAlg, keyType := key.Algorithm(), key.KeyType()
+	if keyAlg != jwa.RS256 {
+		return nil, fmt.Errorf("expected RS256 key, got %s", keyAlg)
+	}
+	if keyType != jwa.RSA {
```

Comment:
I don't think this is needed as `RS256` is already enforces that it is RSA

Reply:
Of course :facepalm: